service/history/workflow/update/update.go

```diff
@@ -231,27 +233,29 @@ func (u *Update) WaitLifecycleStage(
 			return statusAccepted(), nil
 		}
 
-		// If err is coming from user context (user deadline exceeded), then return it to the caller.
+		// If err is coming from the user provided context (context.DeadlineExceeded or context.Canceled), then return it to the caller.
 		if errors.Is(err, ctx.Err()) {
 			metrics.WorkflowExecutionUpdateClientTimeout.With(u.instrumentation.metrics).Record(1)
 			return nil, ctx.Err()
 		}
 
-		// Update uses registryClearedErr when Registry is cleared. This error has special handling here:
-		//   stop waiting for ACCEPTED stage and return Unavailable (retryable) error to the caller.
-		// This error will be retried (by history service handler, or history service client in frontend,
-		// or SDK, or user client). This will recreate Update in the Registry.
-		if errors.Is(err, registryClearedErr) {
-			return nil, WorkflowUpdateAbortedErr
-		}
-
-		// If err is not coming from stCtx,
-		// then it means that the error is from the future itself and needs to be returned to the caller.
-		if !errors.Is(err, stCtx.Err()) {
+		// If err is not coming from stCtx, then it means that the error is coming from the future itself,
+		// If this err is not registryClearedErr, then it needs to be returned to the caller.
+		if !errors.Is(err, stCtx.Err()) && !errors.Is(err, registryClearedErr) {
 			return nil, err
 		}
 	}
 
+	// Because of the checks above err (if is not nil) can be only registryClearedErr here.
+	// It is converted to Unavailable (retryable) error and returned to the caller.
+	// This error will be retried (by history service handler, or history service client in frontend,
+	// or SDK, or user client). This will recreate Update in the Registry.
+	if errors.Is(err, registryClearedErr) {
+		return nil, WorkflowUpdateAbortedErr
+	}
```

Comment:
This is a change. I moved check for `registryClearedErr` out of `if u.accepted.Ready()...` block.